pallets/dapp-staking-v3/src/types.rs

```diff
@@ -588,178 +486,378 @@ where
+        } else {
+            Some(self.staked.period)
+        }
+    }
+
+    /// Earliest era for which the account has staking information or `None` if no staking information exists.
+    pub fn earliest_staked_era(&self) -> Option<EraNumber> {
+        if self.staked.is_empty() {
+            self.staked_future.map(|stake_amount| stake_amount.era)
+        } else {
+            Some(self.staked.era)
+        }
+    }
+
+    /// Cleanup staking information if it has expired.
+    ///
+    /// # Args
+    /// `threshold_period` - last period for which entries can still be considered valid.
+    ///
+    /// `true` if any change was made, `false` otherwise.
+    pub fn maybe_cleanup_expired(&mut self, threshold_period: PeriodNumber) -> bool {
+        match self.staked_period() {
+            Some(staked_period) if staked_period < threshold_period => {
+                self.staked = Default::default();
+                self.staked_future = None;
+                true
+            }
+            _ => false,
+        }
+    }
+
+    /// 'Claim' rewards up to the specified era.
+    /// Returns an iterator over the `(era, amount)` pairs, where `amount`
+    /// describes the staked amount eligible for reward in the appropriate era.
+    ///
+    /// If `period_end` is provided, it's used to determine whether all applicable chunks have been claimed.
+    pub fn claim_up_to_era(
+        &mut self,
+        era: EraNumber,
+        period_end: Option<EraNumber>,
+    ) -> Result<EraStakePairIter, AccountLedgerError> {
+        // Main entry exists, but era isn't 'in history'
+        if !self.staked.is_empty() && era <= self.staked.era {
+            return Err(AccountLedgerError::NothingToClaim);
+        } else if let Some(stake_amount) = self.staked_future {
+            // Future entry exists, but era isn't 'in history'
+            if era < stake_amount.era {
+                return Err(AccountLedgerError::NothingToClaim);
+            }
+        }
+
+        // There are multiple options:
+        // 1. We only have future entry, no current entry
+        // 2. We have both current and future entry
+        // 3. We only have current entry, no future entry
+        let (span, maybe_first) = if let Some(stake_amount) = self.staked_future {
+            if self.staked.is_empty() {
+                ((stake_amount.era, era, stake_amount.total()), None)
+            } else {
+                (
+                    (stake_amount.era, era, stake_amount.total()),
+                    Some((self.staked.era, self.staked.total())),
+                )
+            }
+        } else {
+            ((self.staked.era, era, self.staked.total()), None)
+        };
+
+        let result = EraStakePairIter::new(span, maybe_first);
+
+        // Rollover future to 'current' stake amount
+        if let Some(stake_amount) = self.staked_future.take() {
+            self.staked = stake_amount;
+        }
+        self.staked.era = era.saturating_add(1);
```

Comment:
sometimes using `era + 1` and sometimes `era.saturating_add(1)`. Please make it consistent

Reply:
If I use `era + 1` inside runtime code, then it's wrong 🙂.

I've updated incorrect cases (some can be kept as they're not important like debug_asset and genesis).

Thanks for the good catch!